hydrocarbon_problem/env/env.py

```diff
@@ -46,97 +62,186 @@ def __init__(self, initial_feed_spec: StreamSpecification = DEFAULT_INITIAL_FEED
         timestep_type = dm_env.StepType.MID if not done_overall else dm_env.StepType.LAST
         timestep = dm_env.TimeStep(step_type=timestep_type, observation=observation,
-                                   reward=reward, discount=done)
+                                   reward=reward, discount=discount)
         return timestep
 
 
+    def get_done_overall(self):
+        """Indicates whether the overall episode is complete, because either the
+        maximum number of environment steps have been taken, or the
+        self._stream_numbers_yet_to_be_acted_on is empty. """
+        return self._steps >= self._max_n_steps or \
+                           len(self._stream_numbers_yet_to_be_acted_on) == 0
+
 
-    def _action_to_column_spec(self, action: np.ndarray) -> ColumnInputSpecification:
-        """All actions are assumed to be unbounded and we then translate these into
+
+    def _action_to_column_spec(self, action: Action) -> Tuple[bool, ColumnInputSpecification]:
+        """All actions are assumed to be bounded between -1 and 1, and we then translate these into
         the relevant values for the ColumnInputSpecification."""
-        n_stages = round(np.interp(expit(action[0]), [0, 1], [self._min_n_stages,
-                                                             self._max_n_stages]) + 0.5)
-        feed_stage_location = expit(action[1]) * n_stages
-        reflux_ratio = np.exp(action[2])
-        reboil_ratio = np.exp(action[3])
-        condensor_pressure = np.interp(expit(action[1]), [0, 1], [self._min_pressure,
-                                                                  self._max_pressure])
-        return ColumnInputSpecification(n_stages=n_stages, feed_stage_location=feed_stage_location,
-                                        reflux_ratio=reflux_ratio, reboil_ratio=reboil_ratio,
-                                        condensor_pressure=condensor_pressure)
-
-    def _get_simulated_flowsheet_info(self):
-        """Grabs relevant info from simualted flowsheet."""
+        discrete_action, continuous_action = action
+        n_stages = round(np.interp(continuous_action[0], [-1, 1], self._n_stages_bounds) + 0.5)
+        # feed as fraction between stage 0 and n_stages
+        feed_stage_location = round(np.interp(continuous_action[1], [-1, 1], [0, n_stages]) + 0.5)
+        reflux_ratio = np.interp(continuous_action[2], [-1, 1], self._reflux_ratio_bounds)
+        reboil_ratio = np.interp(continuous_action[3], [-1, 1], self._reflux_ratio_bounds)
+        condensor_pressure = np.interp(continuous_action[4], [-1, 1], self._pressure_bounds)
+        column_spec = ColumnInputSpecification(
+            n_stages=n_stages,
+            feed_stage_location=feed_stage_location,
+            reflux_ratio=reflux_ratio,
+            reboil_ratio=reboil_ratio,
+            condensor_pressure=condensor_pressure)
+        choose_seperate = True if discrete_action == 1 else False
```

Comment:
Is it possible to put line 174 at line 162? So the new ColumnInputSpecification is only set when there is a separation taking place?

Reply:
done